src/detect-http2.c

```diff
@@ -0,0 +1,612 @@
+        return -1;
+    }
+
+    sm->type = DETECT_HTTP2_WINDOW;
+    sm->ctx = (SigMatchCtx *)wu;
+
+    SigMatchAppendSMToList(s, sm, g_http2_match_buffer_id);
+
+    return 0;
+}
+
+/**
+ * \brief this function will free memory associated with uint32_t
+ *
+ * \param ptr pointer to DetectU8Data
+ */
+void DetectHTTP2windowFree(void *ptr)
+{
+    SCFree(ptr);
+}
+
+typedef struct DetectHTTP2settingsSigCtx_ {
+    uint16_t id;   /**identifier*/
+    DetectU32Data *value; /** optional value*/
+} DetectHTTP2settingsSigCtx;
+
+/**
+ * \brief This function is used to match HTTP2 error code rule option on a transaction with those passed via http2.settings:
+ *
+ * \retval 0 no match
+ * \retval 1 match
+ */
+static int DetectHTTP2settingsMatch(DetectEngineThreadCtx *det_ctx,
+                               Flow *f, uint8_t flags, void *state, void *txv, const Signature *s,
+                               const SigMatchCtx *ctx)
+
+{
+    int id = rs_http2_tx_get_settingsid(txv, flags);
+    if (id < 0) {
+        //no settings, no match
+        return 0;
+    }
+
+    const DetectHTTP2settingsSigCtx *setctx = (const DetectHTTP2settingsSigCtx *)ctx;
+    if (setctx->id != id) {
+        return 0;
+    } else if (setctx->value == NULL) {
+        //no value to match
+        return 1;
+    } else {
+        int value = rs_http2_tx_get_settingsvalue(txv, flags);
+        if (value < 0) {
+            return 0;
+        }
+        return DetectU32Match(value, setctx->value);
+    }
+}
+
+static int DetectHTTP2FuncParseSettingsId(const char *str, uint16_t *id)
+{
+    // first parse numeric value
+    if (ByteExtractStringUint16(id, 10, strlen(str), str) >= 0) {
+        return 1;
+    }
+
+    // it it failed so far, parse string value from enumeration
+    int r = rs_http2_parse_settingsid(str);
+    if (r >= 0) {
+        *id = r;
+        return 1;
+    }
+
+    return 0;
+}
+
+#define HTTP2_MAX_SETTINGS_ID_LEN 64
+
+/**
+ * \brief this function is used to attach the parsed http2.settings data into the current signature
+ *
+ * \param de_ctx pointer to the Detection Engine Context
+ * \param s pointer to the Current Signature
+ * \param str pointer to the user provided http2.settings options
+ *
+ * \retval 0 on Success
+ * \retval -1 on Failure
+ */
+static int DetectHTTP2settingsSetup (DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectSignatureSetAppProto(s, ALPROTO_HTTP2) != 0)
+        return -1;
+
+    const char * space = strchr(str, ' ');
+
+    DetectHTTP2settingsSigCtx *http2set = SCCalloc(1, sizeof(DetectHTTP2settingsSigCtx));
+    if (http2set == NULL)
+        return -1;
+
+    if (space) {
+        // a space separates identifier and value
+
+        // copy and isolate first part of string
+        char str_first[HTTP2_MAX_SETTINGS_ID_LEN];
+        if (HTTP2_MAX_SETTINGS_ID_LEN <= space - str) {
+            SCFree(http2set);
+            return -1;
+        }
+        strlcpy(str_first, str, space - str + 1);
```

Comment:
the 3rd arg to strlcpy is the size of the destination buffer, so `sizeof(str_first)`

Reply:
Hmm. I do not want to copy the full buffer, hence the upper check.
`strncpy` was refused by cocci. So I found strlcpy...

And what about Rust for parsing signatures ?

Reply:
No objection to using Rust.

Reply:
Some strategy to advise ? Code strangling ?